### PR TITLE
Chess

### PR DESCRIPTION
**Pull request by TechnoDrive**
Imagine looking at a discord embed.

### CODE REVIEW

**Comment by TechnoDrive** [comment]
I have added 4 chess commands to the bot.
- `.move`
    - If the author is in a chess game, and it's their turn, makes a move for their game, and saves the game state to the database.
- `.challenge`
    - If the author is not in a chess game, create a new game for whoever the author specifies, if none, defaults to themselves. Create a new channel for the game, and then send a message to the author to the channel.
- `.abort`
    - If the author is in a chess game, delete the game channel linked to the game, and then remove the game from the database.
- `.listgames`
    - Fetches all games in the database, and then sends an embed representing them

For these commands to work, I've also added 2 additional npm packages.
- [canvas](https://github.com/Automattic/node-canvas)
    - This package adds the Canvas API to nodejs, which is used in this scenario to generate a chess board image.
- [chess.js](https://github.com/jhlywa/chess.js/blob/master/README.md)
    - This packages is used for move generation / validation, FEN generation, piece movement, board -> array functionality, and much more.
    
Finally, I have created my own class found in `NewChessManager.js`, which provides many helpful methods to interact with the database and chess.js. I have optimized the code, credited code I have taken from other users, and commented the purpose of methods that do not have a clear explanation.

**Comment by pyxld-kris** [comment]
This looks awesome! Just one question from a Discord management point of view.

> * `.challenge`
>   
>   * If the author is not in a chess game, create a new game for whoever the author specifies, if none, defaults to themselves. Create a new channel for the game, and then send a message to the author to the channel.


**Where is this channel created? What are the permissions?**

**Comment by TechnoDrive** [comment]
- **Where is this channel created?**
     - As of right now, it's created on the top of the server, though I am currently looking into creating the channel in a category.
- **What are the permissions?**
    - Anyone can view the channel and send messages to channel. This is intentional however, though at one point I did think about creating roles for the users who are playing, and only let them view the channel. After consideration, I thought there was no harm in letting anyone talk in the channel, as it would only promote discussion towards the game.

**Comment by TechnoDrive** [comment]
I'm pretty sure I can add support for creating the channels in a category, so just tell me if that's something I should do.

**Comment by pyxld-kris** [comment]
Looking at your plugin folder, it looks like we'll have to standardize a few more things for our plugins.

Here's our current structure
 - ### plugin-name
    - **controllers**
       - **commands** _(files representing each chat command this plugin accepts from users)_
       - **events** _(files representing each event that may be triggered and handled by this plugin)_
    - **extends** _(files that extend existing Discord.js classes [TextChannel, Guild, etc], adding functionality and state needed for this plugin)_
    - **structures** _(new classes, or structures, created and used by this specific plugin)_

You've added a few things that don't fit into this mold. For example, the images representing each chess piece. There are also some modules that aren't *classes*, but instead utility functions added. We'll have to handle both of these!

Here's a proposal, let me know what you think:

 - ### plugin-name
    - **assets** _(files like images and JSON which represent discrete pieces of content needed to make this plugin work)_
    - **controllers**
       - **commands** _(files representing each chat command this plugin accepts from users)_
       - **events** _(files representing each event that may be triggered and handled by this plugin)_
    - **extends** _(files that extend existing Discord.js classes [TextChannel, Guild, etc], adding functionality and state needed for this plugin)_
    - **structures** _(new classes, or structures, created and used by this specific plugin)_
    - **utils** _(files with utility modules [functions, classes, etc] which don't fit within **structures**, but are needed fir this specific plugin to function. **Using classes and thinking in terms of Structures is usually preferred!**)_

**Comment by TechnoDrive** [comment]
Everything looks great to me! Should I update my code accordingly?

**Review by pyxld-kris — COMMENTED** [review]
> Everything looks great to me! Should I update my code accordingly?

Yes please! I'm also adding a few comments to specific lines of code with small things

File: src/extensions/Chess/helpers/NewChessManager.js [hunk 1]
```diff
@@ -0,0 +1,289 @@
+module.exports = class {
```
[the remark this answers is not in the document]
Comment (TechnoDrive) [reply src/extensions/Chess/helpers/NewChessManager.js:1]: Alright, I'll fix that as I update the code.

[the remark this answers is not in the document]
Comment (TechnoDrive) [reply src/extensions/Chess/helpers/NewChessManager.js]: Finished.

**Comment by pyxld-kris** [comment]
> * **Where is this channel created?**
>   
>   * As of right now, it's created on the top of the server, though I am currently looking into creating the channel in a category.

How hard is this to change? In our main server, I'm meticulous with the structure of the sidebar, and having a random chess channel at  the very top is something that will most likely drive me crazy.


> * **What are the permissions?**
>   
>   * Anyone can view the channel and send messages to channel. This is intentional however, though at one point I did think about creating roles for the users who are playing, and only let them view the channel. After consideration, I thought there was no harm in letting anyone talk in the channel, as it would only promote discussion towards the game.

There are pros and cons to this. The pro is that a lot more people can see this new feature. The con is that this will most likely send message notifications to anyone who hasn't muted the server, and there will be a LOT of them. How can we handle this?

**Comment by TechnoDrive** [comment]
> How hard is this to change? In our main server, I'm meticulous with the structure of the sidebar, and having a random chess channel at the very top is something that will most likely drive me crazy.
Very easy, I can add support for it right now.

> There are pros and cons to this. The pro is that a lot more people can see this new feature, interact with each other, and potentially up engagement. The con is that this will most likely send message notifications to anyone who hasn't muted the server, and there will be a LOT of them. How can we handle this?
I honestly don't know, but I feel like it's better to just keep it, since most people don't have ping sounds on anyways.

**Comment by guillermosanta** [comment]
It's amazing! Just some things I wanted to point out.

- **Database**
For the database, we are using a Structure called DatabaseHandler, in the .common folder of our plugin folder (currently called extensions). Would it be possible to move all the database calls to the Structure?

- **Embeds**
For sending MessageEmbed, we extended the Discord Message Structure to add a sendEmbed() method, which takes the fields of the MessageEmbed as a parameter. Is that something that can be changed?

**Comment by TechnoDrive** [comment]
Not sure about the database calls yet, I'll have to take a look at it, however, I think I can easily move code to the `sendEmbed()` method.

**Comment by TechnoDrive** [comment]
I have just made it so that only users who possess a "Chess" role can view the game channels.

**Pull request merged**